### PR TITLE
Correct errors due to collection contains constraint

### LINKED ISSUE #2411: And constraint on Has.Member throws
An example

```
var donkey = new object();
var mule = new object();
Assert.That(new[] { donkey }, Has.Member(donkey).And.No.Member(mule));
```
Fails with this exception in 3.8.1

```
System.ArgumentException : The actual value must be an IEnumerable
Parameter name: actual
   at NUnit.Framework.Constraints.SomeItemsConstraint.ApplyTo[TActual](TActual actual) in D:\git\nunit\src\NUnitFramework\framework\Constraints\SomeItemsConstraint.cs:line 64
   at NUnit.Framework.Constraints.NotConstraint.ApplyTo[TActual](TActual actual) in D:\git\nunit\src\NUnitFramework\framework\Constraints\NotConstraint.cs:line 48
   at NUnit.Framework.Constraints.AndConstraint.ApplyTo[TActual](TActual actual) in D:\git\nunit\src\NUnitFramework\framework\Constraints\AndConstraint.cs:line 64
   at NUnit.Framework.Constraints.SomeItemsConstraint.ApplyTo[TActual](TActual actual) in D:\git\nunit\src\NUnitFramework\framework\Constraints\SomeItemsConstraint.cs:line 65
   at NUnit.Framework.Assert.That[TActual](TActual actual, IResolveConstraint expression, String message, Object[] args) in D:\git\nunit\src\NUnitFramework\framework\Assert.That.cs:line 245
   at NUnit.Framework.Assert.That[TActual](TActual actual, IResolveConstraint expression) in D:\git\nunit\src\NUnitFramework\framework\Assert.That.cs:line 228
   at Granta.Services.FunctionalData.Tests.ParameterValueExtraction.ParameterSingleValuesExtractorDatabaseTests.MethodUsingObjectsFailsInDifferentManner() in D:\git\servicelayer\FunctionalDataTests\ParameterValueExtraction\ParameterSingleValuesExtractorDatabaseTests.cs:line 59
```

This is a regression from 3.6 and I suspect was introduced by #2239. 
The right hand constraint of the `And` is getting evaluated against each item of the array instead of against the enumerable. I have tried stepping through and figuring out where things go wrong but without success. 

I *presume* I am using the API correctly - I expected the assertion to check that the array contained `donkey` and did not contain `mule`. If I am wrong, I would appreciate some guidance on how to chain these constraints using the fluent API - or should one write 2 separate assertions?

### LINKED ISSUE #2412: Using fluent syntax unintentionally removed in 3.8
> From @BlythMeister on https://github.com/nunit/nunit/issues/2410
>  with regard to the example code, we had
> Assert.That(myCollection, Contains.Item(expectedItem).Using(new MyCustomComparer()));
> The .Using fluent method existed in 3.7, but was removed in 3.8.
> I was unable to find an equivalent, so changed to
> Assert.That(myCollection.Contains(expectedItem, new MyCustomComparer()));
> Which was a simple enough change as a 1 off, but we had around 100 usages of that old syntax with different collection names, expected names as comparers (across different projects)

This change was unintentional - the removal of CollectionContainsConstraint wasn't intended to cause any breaking changes. 

@mikkelbu did the below analysis of what happened:

> @BlythMeister This is very unfortunate. Actually, @ChrisMaddock spotted the missing Using in aa19bd1 (see #2239 (comment)), so we added Using to the EqualConstraint in bd2091b. However, the entire PR did contain an error, so Contains.Item was again changed in 15c27e5 to return a ContainsConstraint instead of a EqualConstraint. And currently ContainsConstraint does not have a Using method.

@rprouse - will leave you to decide if this should be critical or not, possibly along with https://github.com/nunit/nunit/issues/2411 - I haven't looked into that one.

### PR DESCRIPTION
**Pull request by mikkelbu**
Unfortunately I had to implement a lot of `Using`-overloads
in `SomeItemsConstraint` and add a dynamic check.

Fixes #2411 and fixes #2412

### CODE REVIEW

File: src/NUnitFramework/framework/Constraints/SomeItemsConstraint.cs [hunk 1]
```diff
@@ -67,5 +73,89 @@ public override ConstraintResult ApplyTo<TActual>(TActual actual)
...
+            equalConstraint.Using(comparer);
+            return this;
+        }
+
+        private void CheckPrecondition(string argument)
+        {
+            if (equalConstraint == null)
+            {
+                var message =
+                    "Using can only be called when the underlying constraint is an instance of " + nameof(EqualConstraint);
```
[the remark this answers is not in the document]
Comment (mikkelbu) [reply src/NUnitFramework/framework/Constraints/SomeItemsConstraint.cs]: Looking at the exception messages in the folder src\NUnitFramework\framework\Constraints I think that the the level of detail varies a bit (I've compiled a list of most of them below). Some messages mention concrete .NET classes - or interfaces - others are more high-level (note that the messages are usually about the `actual` parameter to `ApplyTo`, and I guess for that parameter it is easier to relate the exception message to the user code than in our setting, but I could be wrong).

I've no problem with the sentence "Using cannot be used with this combination of constraints", except that it does not communicate how to solve the problem,. But I have a hard time to come up with a better message.

Most of exception messages:
* "Cannot compare using a null reference"
* "A different comparison type was already set."
* "A partial expression may not be resolved"
* "Actual value must be a string"
* "Actual value must be an Exception"
* "The actual value must be a non-null string" + ErrorSubstring
* "Attribute {0} was not found", expectedType
* "Both arguments must be numeric"
* "Cannot check a condition in the past"
* "Cannot create range for a non-numeric value"
* "Neither value implements IComparable or IComparable<T>"
* "Only one directional modifier may be used"
* "Only one Using modifier may be used"
* "Property {0} was not found", name
* "Stream is not readable"
* "Stream is not seekable"
* "The actual value cannot be an empty string"
* "The actual value must be a DirectoryInfo"
* "The actual value must be a string or a non-null IEnumerable or DirectoryInfo"
* "The actual value must be a string" + ErrorSubstring
* "The actual value must be an IDictionary"
* "The actual value must be an IEnumerable"
* "The from value must be less than or equal to the to value."
* "Ulps may only be specified for floating point arguments"
* "Unknown tolerance mode specified"
* "Within modifier may appear only once in a constraint expression"

[the remark this answers is not in the document]
Comment (rprouse) [reply src/NUnitFramework/framework/Constraints/SomeItemsConstraint.cs]: > it does not communicate how to solve the problem,. But I have a hard time to come up with a better message.

I had a hard time coming up with a better message too and was hoping someone would jump in with a suggestions, but nobody did 😦 

Maybe "Using may only be used with constraints that check the equality of the items"? Does this miss any combinations that you can see?

[the remark this answers is not in the document]
Comment (ChrisMaddock) [reply src/NUnitFramework/framework/Constraints/SomeItemsConstraint.cs]: I like your suggestion @rprouse - I think that's a good compromise between enough information, and not being too specific.

> "Using may only be used with constraints that check the equality of the items"

[the remark this answers is not in the document]
Comment (mikkelbu) [reply src/NUnitFramework/framework/Constraints/SomeItemsConstraint.cs]: I also like it (and I cannot come up with anything better). I'll try to do the changes requested tonight

**Comment by ChrisMaddock** [comment]
This looks good to me. @rprouse - given the other changes required for 3.8.2 are already merged in to master, I'm going to do the same here.

**Pull request merged**